Curate charm

lmlg opened the pull request:
In this PR, we adapt the charm to modern practices, write tests (both functional and unit) and update the README file, plus various fixes here and there.

lmlg committed:
Update dependencies

Many required libraries defined in test-requirements.txt,
requirements.txt and wheelhouse were unneeded and/or used
deprecated versions.

lmlg committed:
Update tox.ini file

Add the 'build' rule and also update the needed dependencies
for tests to properly run.

lmlg committed:
Make the implementation file PEP8 compliant

lmlg committed:
Write unit tests

These set of tests will check for various configuration settings,
since the charm is mostly configuration-only.

lmlg committed:
Add .gitignore file

lmlg committed:
Add copyright and metadata files

Commit:
Modify YAML test files

The changes in the layer file is needed so that tests are included
in the build directory, whereas the modifications for the bundle
file set the proper variables to connect to NetApp.

lmlg committed:
Update unit tests

lmlg committed:
Add the bundles and overlay for the functional tests

This charm requires an overlay to properly set the access
and credentials to connect to a NetApp instance. Additionally,
add the functional bundles following the blueprint set by
the cinder charm.

lmlg committed:
Update the tests.yaml file with the new functional tests bundles

lmlg committed:
Standardize functional tests bundles

Make the bundles more in line with what's already being used in
the cinder charm. Also, add a few variables that don't belong in
the overlay back in the YAML files.

lmlg committed:
Use the expected name for the overlay file

Instead of using a non-standard name and have the user specify
the overlay by hand, make it so that 'tox -e func' can pick it
up by virtue of using a specific name.

lmlg committed:
Add the .zuul and osci YAML files needed for CI

lmlg committed:
Remove the .zuul and osci files

This is needed in order for the opendev/project-config pipeline
to pass for this branch.

lourot commented on copyright:
```diff
@@ -0,0 +1,17 @@
+
+Files: *
+Copyright: Copyright 2021, Canonical Ltd., All Rights Reserved.
+License: GPL-3
```
Do we really want this license? AFAIK we prefer the Apache license, e.g. [here](https://github.com/openstack/charm-keystone/blob/master/copyright).

lourot commented on metadata.yaml:
```diff
@@ -0,0 +1 @@
+src/metadata.yaml
```
If this is a reactive charm, you don't need to have a `metadata.yaml` file in the root directory. Only under `src/` so you could remove this symlink.

lmlg replied:
OK, but `charm-proof` fails if this file doesn't exist. Should I drop the call to `charm-proof` for the `pep8` environment ?

lourot commented on requirements.txt:
```diff
@@ -1,21 +1,3 @@
-# This file is managed centrally by release-tools and should not be modified
```
If this is a reactive charm, you have to keep that disclaimer and follow it, i.e. you're not allowed to make special modifications to this file. Any change you need needs to go to https://github.com/openstack-charmers/release-tools/blob/master/global/source-zaza/requirements.txt first. We want to avoid charms from becoming snowflakes, and want to keep the boilerplate of our 60+ OpenStack charms the same. This file can only be a verbatim copy of what is in `release-tools`.

lourot commented on src/layer.yaml:
```diff
@@ -1,24 +1,12 @@
-"includes":
-- "layer:options"
-- "layer:basic"
-- "interface:tls-certificates"
```
don't we need this interface anymore for example? Are we getting it indirectly through one of the layers you kept?

lmlg replied:
I'm basing this file off the cinder-purestorage charm, which I think is the one that resembles this one the most.

lourot replied:
but maybe we don't need TLS for purestorage but we do for netapp? I don't know the answer, just checking we don't remove things by accident.

lmlg replied:
I believe that's unlikely, as this charm is 'configuration-only' - i.e: It only provides an interface for `cinder_configuration` and does nothing beyond that.

lourot commented on src/layer.yaml:
```diff
@@ -1,24 +1,12 @@
-  ".gitignore", ".gitreview", ".zuul.yaml", "README.md", "requirements.txt"]
-"options":
-  "basic":
-    "use_venv": !!bool "true"
```
I think these options are for `layer:basic`. Are you really really sure this is the right move to remove this option for example? Removing this option may prevent the charm from installing properly on some Ubuntu series if I remember correctly.

lourot replied:
I see, [`true` is the default](https://github.com/juju-solutions/layer-basic/blob/master/layer.yaml) already

lourot commented on test-requirements.txt:
```diff
@@ -1,18 +1,7 @@
-# This file is managed centrally by release-tools and should not be modified
```
Same here, you have to respect this disclaimer. Don't turn this charm into a snowflake please :)

lourot commented on tox.ini:
```diff
@@ -14,6 +14,8 @@ skipsdist = True
 sitepackages = False
 # NOTE: Avoid false positives by not skipping missing interpreters.
 skip_missing_interpreters = False
+requires = pip < 20.3
```
Re: the disclaimer at the top of the file. Is it now aligned with `release-tools` or are you diverging from it?

lourot reviewed (changes requested):
Thanks a lot. I reviewed only the boilerplate so far, not the actual charm and test implementations.

lmlg committed:
Copy over files from release-tools

Instead of doing piecemeal modifications to the old files of this
charm, use the readily available files from release-tools.

lmlg committed:
Fix test class path

lmlg committed:
Streamline the layer.yaml file

Base it off the cinder-purestorage one.

lmlg committed:
Remove the .zuul file

lmlg committed:
Update the README files

lourot commented on README.md:
```diff
@@ -1,22 +0,0 @@
-# Overview
```
For this file, as we do [in some other reactive charms](https://github.com/openstack/charm-ironic-api), the best would be to just do a symlink to `src/README.md`

lourot commented on .gitignore:
```diff
@@ -0,0 +1,9 @@
+build
+layers
+.tox
```
Also in a separate review let's delete this folder from git (it has been checked in by mistake). But please not in this review because this will delete about 5000 files and it will be impossible to review the interesting bits :D

lourot commented on src/README.md:
```diff
@@ -1,19 +1,101 @@
-See config.yaml for details of configuration options.
+### `netapp-storage-family`
+
+The storage family type that is used for the storage system. Can be "ontap_cluster" for clustering data ONTAP, or "eseries".
```
Nitpick, I think it's better to use `backticks` for values rather than "double-quotes". It renders better on the charmstore.

```
better `backticks` than "double-quotes"
```

Note that on the [charm-guide](https://docs.openstack.org/charm-guide/latest/2110.html) we seem to also be using sometimes 'single-quotes' for values, but never "double-quotes"

lourot commented on src/README.md:
```diff
@@ -1,19 +1,101 @@
+
+This charm's primary use is as a backend for the cinder charm. To do so, add a relation betweeen both charms:
+
+  juju add-relation cinder-netapp:storage-backend cinder:storage-backend
```
If you want this line to be rendered as a

    code block

you need to make it start with at least 4 spaces

lourot commented on src/README.md:
```diff
@@ -1,19 +1,101 @@
+
+[cg]: https://docs.openstack.org/charm-guide
+[cdg]: https://docs.openstack.org/project-deploy-guide/charm-deployment-guide
+[lp-bugs-charm-cinder-netapp]: https://bugs.launchpad.net/charm-cinder-netapp/+filebug
```
This doesn't seem to exist yet

lmlg replied:
Yeah, I added this so that it's already in place for when the charm is incorporated into opendev.

lourot commented on src/tests/bundles/focal-ussuri.yaml:
```diff
@@ -1,59 +1,75 @@
-  mysql:
-    charm: cs:~openstack-charmers-next/percona-cluster
+  mysql-innodb-cluster:
+    charm: cs:~openstack-charmers/mysql-innodb-cluster
```
You have removed `-next` everywhere and I believe this is not what you want

lourot commented on src/tests/tests.yaml:
```diff
@@ -1,9 +1,17 @@
+  - focal-ussuri
+  - focal-victoria
+  - groovy-victoria
+  - hirsute-wallaby
```
Note that [we are in the process of adding also focal-xena and impish-xena to all charms](https://review.opendev.org/q/topic:xena-bundles) so you probably want to do the same for this charm as well

lourot reviewed (changes requested):
I have now reviewed everything, thanks a lot, LGTM in general

lmlg committed:
Use backquotes instead of double quotes in the README file

lmlg committed:
Add a symlink to src/README file in top directory

lmlg committed:
Update test bundles

lmlg committed:
Add bundles for focal-xena and impish-xena

lmlg committed:
Add the focal-wallaby bundle and update the tests.yaml file

lourot commented on src/tests/tests.yaml:
```diff
@@ -1,9 +1,23 @@
-  - xenial-stein
+  - bionic-ussuri
+dev_bundles:
+  - impish-xena
```
Should we add `focal-xena` to this list as well?

lmlg replied:
Yep, we should.

lourot reviewed (changes requested):
All my comments have been addressed, thanks a lot! Only one nitpick left and I'm happy to land this.

lmlg committed:
Add missing 'focal-xena' bundle in tests.yaml
